spec/02-integration/05-proxy/19-grpc_proxy_spec.lua

```diff
@@ -45,373 +46,471 @@ end
+              ["Content-Type"] = "application/grpc"
+            }
+          }))
+          local body = assert.res_status(426, res)
+          local json = cjson.decode(body)
+          assert.equal("Please use HTTP2 protocol", json.message)
+        end)
+
+        it("non-grpc request on grpc route (no content-type)", function()
+          local body, headers = proxy_client_h2c({
+            headers = {
+              ["method"] = "POST",
+              [":authority"] = "grpc",
+            }
+          })
+          local json = cjson.decode(body)
+          assert.same("415", headers:get(":status"))
+          assert.same("Non-gRPC request matched gRPC route", json.message)
+        end)
+
+        it("non-grpc request on grpcs route (no content-type)", function()
+          local body, headers = proxy_client_h2({
+            headers = {
+              ["method"] = "POST",
+              [":authority"] = "grpcs",
+            }
+          })
+          local json = cjson.decode(body)
+          assert.same("415", headers:get(":status"))
+          assert.same("Non-gRPC request matched gRPC route", json.message)
+        end)
+
+        it("non-grpc request on grpc route (non-grpc content-type)", function()
+          local body, headers = proxy_client_h2c({
+            headers = {
+              ["method"] = "POST",
+              ["content-type"] = "application/json",
+              [":authority"] = "grpc",
+            }
+          })
+          local json = cjson.decode(body)
+          assert.same("415", headers:get(":status"))
+          assert.same("Non-gRPC request matched gRPC route", json.message)
+        end)
+
+        it("non-grpc request on grpcs route (non-grpc content-type)", function()
+          local body, headers = proxy_client_h2({
+            headers = {
+              ["method"] = "POST",
+              ["content-type"] = "application/json",
+              [":authority"] = "grpcs",
+            }
+          })
+          local json = cjson.decode(body)
+          assert.same("415", headers:get(":status"))
+          assert.same("Non-gRPC request matched gRPC route", json.message)
+        end)
+
+        it("grpc on grpcs route", function()
+          local ok, resp = proxy_client_grpc({
+            service = "hello.HelloService.SayHello",
+            body = {
+              greeting = "world!"
+            },
+            opts = {
+              ["-authority"] = "grpcs",
+            }
+          })
+          assert.falsy(ok)
+
+          if flavor == "expressions" then
+            assert.matches("Code: NotFound", resp, nil, true)
+            assert.matches("Message: NotFound", resp, nil, true)
+
+          else
+            assert.matches("Code: Canceled", resp, nil, true)
+            assert.matches("Message: gRPC request matched gRPCs route", resp, nil, true)
+          end
+        end)
+      end)
+      
+      if strategy ~= "off" then
+        describe("grpcs with tls", function()
```

Comment:
We need to add lousy cases for this test and use the wrong ca_certificate and cert for the bad case. You'll be able to follow this to do it. https://github.com/Kong/kong/blob/master/spec/02-integration/05-proxy/18-upstream_tls_spec.lua#L350